synapse/federation/sender/__init__.py

```diff
@@ -337,34 +341,84 @@ async def handle_event(event: EventBase) -> None:
+                        return room_id, {
+                            event: dests for event, dests in handled_events
+                        }
 
                 events_by_room = {}  # type: Dict[str, List[EventBase]]
                 for event in events:
                     events_by_room.setdefault(event.room_id, []).append(event)
 
-                await make_deferred_yieldable(
+                per_room_events_and_dests = await make_deferred_yieldable(
                     defer.gatherResults(
                         [
-                            run_in_background(handle_room_events, evs)
-                            for evs in events_by_room.values()
+                            run_in_background(handle_room_events, rid, evs)
+                            for rid, evs in events_by_room.items()
                         ],
                         consumeErrors=True,
                     )
+                )  # type: List[Tuple[str, Dict[EventBase, Collection[str]]]]
+
+                # Tuples of room_id + destination to their max-seen stream_ordering
+                room_with_dest_stream_ordering = {}  # type: Dict[Tuple[str, str], int]
+
+                # List of events to send to each destination
+                event_to_dests = {}  # type: Dict[str, List[EventBase]]
+
+                # For each roomID + events with destinations pair...
+                for room_id, event_and_dests in per_room_events_and_dests:
+                    # ...get every event with their destinations...
+                    for event, destinations in event_and_dests.items():
+
+                        # (we got this from the database, it's filled)
+                        assert event.internal_metadata.stream_ordering
+
+                        # ...iterate over those destinations..
+                        for destination in destinations:
+                            if (
+                                room_with_dest_stream_ordering.setdefault(
+                                    (room_id, destination), 0
+                                )
+                                < event.internal_metadata.stream_ordering
+                            ):
+                                # ...update their stream-ordering...
+                                room_with_dest_stream_ordering[
+                                    (room_id, destination)
+                                ] = event.internal_metadata.stream_ordering
+
+                            # ...and add the event to each destination queue.
+                            event_to_dests.setdefault(destination, []).append(event)
+
+                # Bulk-store destination_rooms stream_ids
+                await self.store.bulk_store_destination_rooms_entries(
+                    room_with_dest_stream_ordering
                 )
 
+                # Send corresponding events to each destination queue
+                self._distribute_pdus(event_to_dests)
```

Comment:
distribution happens *after* database access, because then a database-access-crash-loop will cause no events to be spammed to other servers, but a crash loop will also cause federation to stall, which is much more noticeable in the logs (as only the errors will be visible)

also, i think it's "safer" to *not* have a race condition where `destination_room` values are set after pdus are actually distributed

anywho, if either of these functions throw, it wont reach `update_federation_out_pos`, and it'll loop around with `finally:` and the next poke from a database sequence